rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala

```diff
@@ -283,12 +291,11 @@ object Substitute {
               newLocallyFree = locallyFree.until(env.shift)
             } yield Expr(exprInstance = ETupleBody(ETuple(pss, newLocallyFree, connectiveUsed)))
 
-          case ESetBody(ESet(ps, locallyFree, connectiveUsed)) =>
+          case ESetBody(ParSet(SortedHashSet(ps), connectiveUsed)) =>
             for {
               pss <- ps.toList
                       .traverse(p => s1(p))
-              newLocallyFree = locallyFree.until(env.shift)
-            } yield Expr(exprInstance = ESetBody(ESet(pss, newLocallyFree, connectiveUsed)))
+            } yield Expr(exprInstance = ESetBody(ParSet(SortedHashSet(pss.toSeq), connectiveUsed)))
```

Comment:
@guardbotmk3 I am not sure about this change. Is calculating `locallyFree`s based on what `Par`s go into the `Set` yield the same `locallyFree` as `locallyFree.until(env.shift)` ? 

Can you explain what this _shifting_ means?

Reply:
You can allow it to lazily re-calculate, but there's an efficiency loss, because if it is cached, we can compute the updated value in O(1) time, vs O(n) if we just invalidate it.
I'm not worried about it at the moment.

Reply:
I refactored the `locallyFree` to be a `Coeval` now. Please take a look @guardbotmk3